File: packages/forge-core/src/forge_core/domain/entities/card.py
from pydantic import BaseModel, Field

from .enums import CardClass, CardType, Faction


class GlobalEffect(BaseModel):
    target_class: CardClass | None = None
    target_faction: Faction | None = None
    multiplier: float


class PassiveEffect(BaseModel):
    type: str
    value: float


class ActiveEffect(BaseModel):
    type: str
    uses_per_battle: int


class TroopCardDefinition(BaseModel):
    definition_id: str
    card_type: CardType = CardType.TROOP
    name: str
    faction: Faction
    card_class: CardClass
    base_power: int


class HeroCardDefinition(BaseModel):
    definition_id: str
    card_type: CardType = CardType.HERO
    name: str
    faction: Faction
    card_class: CardClass
    base_power: int
    max_deployments: int


class GeneralCardDefinition(BaseModel):
    definition_id: str
    card_type: CardType = CardType.GENERAL
    name: str
    faction: Faction
    card_class: CardClass
    base_power: int
    global_effect: GlobalEffect


class RelicCardDefinition(BaseModel):
    definition_id: str
    card_type: CardType = CardType.RELIC
    name: str
    passive_effect: PassiveEffect | None = None
    active_effect: ActiveEffect | None = None


class CardInstance(BaseModel):
    instance_id: str
    definition_id: str
    level: int = 1
    experience: int = 0


class HeroCardInstance(BaseModel):
    instance_id: str
    definition_id: str
    deployments_remaining: int

File: packages/ts/shared-schema/src/entities/card.ts
import { Faction } from '../enums/faction';
import { CardClass } from '../enums/card-class';
import { CardType } from '../enums/card-type';

/** Base card definition — shared attributes across all card types */
export interface CardDefinition {
  definitionId: string;
  cardType: CardType;
  name: string;
  faction: Faction;
  cardClass: CardClass;
  basePower: number;
}

/** A troop card definition */
export interface TroopCardDefinition extends CardDefinition {
  cardType: CardType.TROOP;
}

/** A hero card definition */
export interface HeroCardDefinition extends CardDefinition {
  cardType: CardType.HERO;
  maxDeployments: number;
}

/** A general card definition */
export interface GeneralCardDefinition extends CardDefinition {
  cardType: CardType.GENERAL;
  globalEffect: GlobalEffect;
}

/** A relic card definition */
export interface RelicCardDefinition {
  definitionId: string;
  cardType: CardType.RELIC;
  name: string;
  passiveEffect: PassiveEffect | null;
  activeEffect: ActiveEffect | null;
}

/** Effect applied globally by a general */
export interface GlobalEffect {
  targetClass: CardClass | null;
  targetFaction: Faction | null;
  multiplier: number;
}

/** Passive relic effect active for the entire battle */
export interface PassiveEffect {
  type: string;
  value: number;
}

/** Active relic effect with limited uses per battle */
export interface ActiveEffect {
  type: string;
  usesPerBattle: number;
}

/** A card instance — a specific copy of a card in a player's collection */
export interface CardInstance {
  instanceId: string;
  definitionId: string;
  level: number;
  experience: number;
}

/** A hero card instance with deployment tracking */
export interface HeroCardInstance {
  instanceId: string;
  definitionId: string;
  deploymentsRemaining: number;
}
